### ISSUE TITLE
Errors when trying to sign large numbers of files

### ISSUE THREAD

**Issue by aeisenberg** [Azure/azure-code-signing-action#21]
We are using this action to sign our binaries. We are signing lots (~120) binaries at a time. On a fairly regular basis, we get errors. 

Some errors look like this (but for seemingly arbitrary files):

```
SignTool Error: The specified timestamp server either could not be reached or
returned an invalid response.

SignTool Error: An error occurred while attempting to sign: D:\a\<PATH_REDACTED>
```

Other errors are a timeout, like this:

```
Submitting digest for signing...
Unhandled managed exception
System.AggregateException: Retry failed after 4 tries. Retry settings can be adjusted in ClientOptions.Retry. (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)
 ---> System.Threading.Tasks.TaskCanceledException: The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Security.SslStream.EnsureFullTlsFrameAsync[TIOAdapter](TIOAdapter adapter)
   at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](TIOAdapter adapter, Memory`1 buffer)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpConnection.InitialFillAsync(Boolean async)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   --- End of inner exception stack trace ---
   at Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
SignTool Error: An unexpected internal error has occurred.
   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)<---

 ---> (Inner Exception #3) System.Threading.Tasks.TaskCanceledException: The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Security.SslStream.EnsureFullTlsFrameAsync[TIOAdapter](TIOAdapter adapter)
   at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](TIOAdapter adapter, Memory`1 buffer)
   at System.Net.Http.HttpConnection.InitialFillAsync(Boolean async)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, Boolean async)
   at Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   --- End of inner exception stack trace ---
   at Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
   at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)<---
```

The timeout looks like it was tried 4 times and timed out each time. The first error does not appear to have been retried.

**Comment by japarson** [comment]
I've created a separate issue for the timestamp errors. We can add a system to help fix that.

The timeout errors have happened during AAD token authentication:
`Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(.....)`

Which authentication method are you using?

**Comment by aeisenberg** [comment]
To be honest, I'm not exactly sure. I am using the following secrets and these were created for me by some internal tooling. If it would be helpful, I can do some digging to figure out how it works.

```
          azure-tenant-id: ${{ secrets.SPN_CODEQL_CODE_SIGNING_TENANT_ID }}
          azure-client-id: ${{ secrets.SPN_CODEQL_CODE_SIGNING_CLIENT_ID }}
          azure-client-secret: ${{ secrets.SPN_CODEQL_CODE_SIGNING }}
```

**Comment by japarson** [comment]
In that case, you are using an [EnvironmentCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredential?view=azure-dotnet). I'm communicating with the Azure Code Signing team now to figure out what might be going wrong.

At what point in the process does this timeout error usually occur? Is it, for example, at the beginning of the run? Or does it happen at some point later after hundreds of files have already been signed?

**Comment by aeisenberg** [comment]
It seems to be happening deep into the signing job, so not at the beginning.

**Comment by aeisenberg** [comment]
Hmmmm...it really seems all over the map.

Successful jobs take as much as 38 minutes to run or as little as 7 minutes. For these runs, we are splitting the files to sign into 4 chunks and signing them serially. For the 38 minute job, the longest signing step is taking almost 13 minutes. This is strange since I had the timeout set to 1500 seconds.



Here's something interesting I just found in the 38 minute job, an operation that took over 100s (all the other operations took < 1s).

```
Wed, 17 Jan 2024 09:25:03 GMT Submitting digest for signing...
Wed, 17 Jan 2024 09:25:04 GMT
Wed, 17 Jan 2024 09:25:04 GMT OperationId e33b5a6e-54b9-4c1e-b375-09d4d1bc76bb: InProgress
Wed, 17 Jan 2024 09:26:46 GMT
Wed, 17 Jan 2024 09:26:46 GMT Signing completed with status 'Succeeded' in 102.9866863s
```

Then, for the failing jobs (at least the jobs failing due to timeouts and retries), I see:

```
Wed, 17 Jan 2024 15:42:43 GMT Submitting digest for signing...
Wed, 17 Jan 2024 15:49:29 GMT Unhandled managed exception
Wed, 17 Jan 2024 15:49:29 GMT System.AggregateException: Retry failed after 4 tries. Retry settings can be adjusted in ClientOptions.Retry. (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)
```

Notice the 7 minute delay. I guess this can be explained by 4 retries each with a timeout of 1:40. But, why is it timing out? I would be surprised if this has to do with the credentials.

**Comment by japarson** [comment]
Hi @aeisenberg, still investigating this issue. It's possible there are some network issues happening on the host machine. Do you use the GitHub-hosted runners for this Action or self-hosted?

Would you be able to send me an email <japarson@microsoft.com> that contains your Azure Code Signing account name, certificate profile name, and subscription ID? This is so that I can pass the info off to the ACS team for investigation.

Could you try excluding every credential except for `EnvironmentCredential`? This might improve performance and help us isolate the problem. It might also be helpful if you could send us the full logs from this run (if it fails after disabling the credentials).

**Comment by aeisenberg** [comment]
None of those env vars are set. What other credentials might there be?

**Comment by japarson** [comment]
The Action performs authentication using [DefaultAzureCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet) which tries a bunch of authentication methods in order - moving on to the next one if one fails.

Each authentication method can be disabled individually so that we don't waste any time trying it when it's never going to pass. See the docs here: https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet

In your case, since you are using the `EnvironmentCredential` specifically, you can use the following inputs:
```C#
exclude-environment-credential: false
exclude-managed-identity-credential: true
exclude-shared-token-cache-credential: true
exclude-visual-studio-credential: true
exclude-visual-studio-code-credential: true
exclude-azure-cli-credential: true
exclude-azure-powershell-credential: true
exclude-interactive-browser-credential: true
```

This could theoretically make you fail faster if for some reason the `EnvironmentCredential` fails and can also help us ignore irrelevant errors from authentication methods that you aren't using.

**Comment by aeisenberg** [comment]
And, we are using GitHub hosted runners (`windows-latest`).

**Comment by japarson** [comment]
For how long has this error been happening for you?
> System.AggregateException: Retry failed after 4 tries. Retry settings can be adjusted in ClientOptions.Retry. (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)

Did it appear suddenly at some point in time or has it always been an intermittent issue?

**Comment by aeisenberg** [comment]
We switched from our previous signing provider a week ago and saw these problems immediately.

**Comment by aeisenberg** [comment]
OK. Just had the error above. I'll send you the logs directly.

**Comment by japarson** [comment]
1. Regarding the signing operations that take ~100 seconds (Wed, 17 Jan 2024 09:26:46 GMT Signing completed with status 'Succeeded' in 102.9866863s), this is apparently a known issue with the WUS region where ~10% of signing requests will be very slow. The ACS team says a possible solution would be to re-route your account to the EUS region. Is that something you'd be interested in?
2. Can you check if the System.AggregateException still occurs if you don't use any timestamp server?
3. Are you running any other jobs/steps/actions in parallel with the signing action? Any reason to suspect some other software is interfering with the action?

**Comment by aeisenberg** [comment]
1. I have no problems using the EUS server. I'm guessing doing this would be just using this url in the `endpoint` property: https://eus.codesigning.azure.net? I'll try that next.
2. Maybe I'll hold off on this and try one thing at a time. Interestingly, the `AggregateException` stack trace message says: "The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout". Is it possible that my other gaps in time are when the timeouts occur on the first or second try, but things succeed in the third try?
3. We are using a standard actions runner and nothing is else is running except for standard OS processes. I'd be surprised if this were the issue.

**Comment by aeisenberg** [comment]
The switch to EUS seems to have fixed the problem. Build times are down to ~15-18 minutes and I haven't seen a timeout yet. Previously, build times were 18-45 min with timeouts every 3-5 runs.

From my perspective, this issue is solved, but I do wonder why the WUS server has these timeouts and if this is something that will be fixed in the future.

**Comment by japarson** [comment]
Glad to hear this fixed the problem. I'll leave this issue open until I add some documentation about "Known Issues". The Azure Code Signing team is currently working with Azure to solve the WUS problem.
